server/src/main/scala/org/http4s/server/middleware/CSRF.scala

```diff
@@ -244,55 +239,156 @@ final class CSRF[F[_], G[_]] private[middleware] (
+    def withClock(clock: Clock): CSRFBuilder[F, G]                                 = copy(clock = clock)
+    def withOnFailure(onFailure: Response[G]): CSRFBuilder[F, G]                   = copy(onFailure = onFailure)
+    def withCreateIfNotFound(createIfNotFound: Boolean): CSRFBuilder[F, G]         = copy(createIfNotFound = createIfNotFound)
+    def withKey(key: SecretKey): CSRFBuilder[F, G]                                 = copy(key = key)
+    def withHeaderCheck(headerCheck: Request[G] => Boolean): CSRFBuilder[F, G]     = copy(headerCheck = headerCheck)
+    def withCSRFCheck(csrfCheck: CSRF[F, G] => CSRFCheck[F, G]): CSRFBuilder[F, G] = copy(csrfCheck = csrfCheck)
+
+    private def cookieMod(f: CookieSettings => CookieSettings)            = copy(cookieSettings = f(cookieSettings))
+
+    def withCookieName(cookieName: String): CSRFBuilder[F, G]             = cookieMod(_.copy(cookieName = cookieName))
+    def withCookieSecure(secure: Boolean): CSRFBuilder[F, G]              = cookieMod(_.copy(secure = secure))
+    def withCookieHttpOnly(httpOnly: Boolean): CSRFBuilder[F, G]          = cookieMod(_.copy(httpOnly = httpOnly))
+    def withCookieDomain(domain: Option[String]): CSRFBuilder[F, G]       = cookieMod(_.copy(domain = domain))
+    def withCookiePath(path: Option[String]): CSRFBuilder[F, G]           = cookieMod(_.copy(path = path))
+    def withCookieExtension(extension: Option[String]): CSRFBuilder[F, G] = cookieMod(_.copy(extension = extension))
+
+    def build: CSRF[F, G] =
+      new CSRF[F, G](headerName, cookieSettings, clock, onFailure, createIfNotFound, key, headerCheck, csrfCheck)
+
+  }
+
+  private[middleware] final case class CookieSettings(
+    cookieName: String,
+    secure: Boolean,
+    httpOnly: Boolean,
+    domain: Option[String] = None,
+    path: Option[String] = None,
+    extension: Option[String] = None
+  )
+
+  ///
+
+  type CSRFCheck[F[_], G[_]] = (Request[G], F[Response[G]]) => F[Response[G]]
+
+  def checkCSRFDefault[F[_], G[_]](implicit F: Sync[F]): CSRF[F, G] => CSRFCheck[F, G] =
+    csrf => (r, http) => csrf.getHeaderToken(r).fold(csrf.onfailureF)(csrf.checkCSRFToken(r, http, _))
+
+  def checkCSRFinHeaderAndForm[F[_], G[_]: Sync](fieldName: String, nt: G ~> F)(
+    implicit F: Sync[F]
+  ): CSRF[F, G] => CSRFCheck[F, G] = { csrf => (r, http) =>
+
+    def getFormToken: F[Option[String]] = {
+
+      def extractToken: G[Option[String]] =
+        r.attemptAs[UrlForm].value.map(_.fold(_ => none[String], _.values.get(fieldName).flatMap(_.headOption)))
+
+      r.headers.get(headers.`Content-Type`) match {
+        case Some(headers.`Content-Type`(MediaType.application.`x-www-form-urlencoded`, _)) => nt(extractToken)
+        case _                                                                              => F.pure(none[String])
+      }
+    }
+
+    for {
+      fst <- F.pure(csrf.getHeaderToken(r))
```

Comment:
Why is the `pure` necessary here, seems this should just be on the next line

Reply:
Not sure why I did that, perhaps because a fold resulted in `v => Some(v)` - If you have a more elegant approach please share:) on the phone now

Reply:
You're right, all of them I know of would involve at least one thunk. Here's hoping that someone else has a clean idea.